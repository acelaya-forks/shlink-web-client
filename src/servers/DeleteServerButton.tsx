import { faMinusCircle as deleteIcon } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { useToggle } from '@shlinkio/shlink-frontend-kit';
import { clsx } from 'clsx';
import type { FC, PropsWithChildren } from 'react';
import { useCallback } from 'react';
import { useNavigate } from 'react-router';
import type { FCWithDeps } from '../container/utils';
import { componentFactory, useDependencies } from '../container/utils';
import type { ServerWithId } from './data';
import type { DeleteServerModalProps } from './DeleteServerModal';

export type DeleteServerButtonProps = PropsWithChildren<{
  server: ServerWithId;
  className?: string;
  textClassName?: string;
}>;

type DeleteServerButtonDeps = {
  DeleteServerModal: FC<DeleteServerModalProps>;
};

const DeleteServerButton: FCWithDeps<DeleteServerButtonProps, DeleteServerButtonDeps> = (
  { server, className, children, textClassName },
) => {
  const { DeleteServerModal } = useDependencies(DeleteServerButton);
  const [isModalOpen, , showModal, hideModal] = useToggle();
  const navigate = useNavigate();
  const onClose = useCallback((confirmed: boolean) => {
    hideModal();
    if (confirmed) {
      navigate('/');
    }
  }, [hideModal, navigate]);

  return (
    <>
      <button type="button" className={clsx(className, 'p-0 bg-transparent border-0')} onClick={showModal}>
        {!children && <FontAwesomeIcon fixedWidth icon={deleteIcon} />}
        <span className={textClassName}>{children ?? 'Remove this server'}</span>
      </button>

      <DeleteServerModal server={server} open={isModalOpen} onClose={onClose} />
    </>
  );
};

export const DeleteServerButtonFactory = componentFactory(DeleteServerButton, ['DeleteServerModal']);
