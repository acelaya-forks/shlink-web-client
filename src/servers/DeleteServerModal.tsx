import { CardModal } from '@shlinkio/shlink-frontend-kit/tailwind';
import type { FC } from 'react';
import { useCallback } from 'react';
import type { ServerWithId } from './data';

export type DeleteServerModalProps = {
  server: ServerWithId;
  onClose: (confirmed: boolean) => void;
  open: boolean;
};

type DeleteServerModalConnectProps = DeleteServerModalProps & {
  deleteServer: (server: ServerWithId) => void;
};

export const DeleteServerModal: FC<DeleteServerModalConnectProps> = ({ server, onClose, open, deleteServer }) => {
  const onConfirm = useCallback(() => {
    deleteServer(server);
    onClose(true);
  }, [deleteServer, onClose, server]);

  return (
    <CardModal
      open={open}
      title="Remove server"
      variant="danger"
      onClose={() => onClose(false)}
      onConfirm={onConfirm}
      confirmText="Delete"
    >
      <p>Are you sure you want to remove <b>{server ? server.name : ''}</b>?</p>
      <p>
        <i>
          No data will be deleted, only the access to this server will be removed from this device.
          You can create it again at any moment.
        </i>
      </p>
    </CardModal>
  );
};
