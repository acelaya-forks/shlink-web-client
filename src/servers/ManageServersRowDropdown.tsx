import { faCircle as toggleOnIcon } from '@fortawesome/free-regular-svg-icons';
import {
  faBan as toggleOffIcon,
  faEdit as editIcon,
  faMinusCircle as deleteIcon,
  faPlug as connectIcon,
} from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { RowDropdownBtn, useToggle } from '@shlinkio/shlink-frontend-kit';
import type { FC } from 'react';
import { Link } from 'react-router';
import { DropdownItem } from 'reactstrap';
import type { FCWithDeps } from '../container/utils';
import { componentFactory, useDependencies } from '../container/utils';
import type { ServerWithId } from './data';
import type { DeleteServerModalProps } from './DeleteServerModal';

export type ManageServersRowDropdownProps = {
  server: ServerWithId;
};

type ManageServersRowDropdownConnectProps = ManageServersRowDropdownProps & {
  setAutoConnect: (server: ServerWithId, autoConnect: boolean) => void;
};

type ManageServersRowDropdownDeps = {
  DeleteServerModal: FC<DeleteServerModalProps>
};

const ManageServersRowDropdown: FCWithDeps<ManageServersRowDropdownConnectProps, ManageServersRowDropdownDeps> = (
  { server, setAutoConnect },
) => {
  const { DeleteServerModal } = useDependencies(ManageServersRowDropdown);
  const [isModalOpen,, showModal, hideModal] = useToggle();
  const serverUrl = `/server/${server.id}`;
  const { autoConnect: isAutoConnect } = server;
  const autoConnectIcon = isAutoConnect ? toggleOffIcon : toggleOnIcon;

  return (
    <RowDropdownBtn minWidth={isAutoConnect ? 210 : 170}>
      <DropdownItem tag={Link} to={serverUrl}>
        <FontAwesomeIcon icon={connectIcon} fixedWidth /> Connect
      </DropdownItem>
      <DropdownItem tag={Link} to={`${serverUrl}/edit`}>
        <FontAwesomeIcon icon={editIcon} fixedWidth /> Edit server
      </DropdownItem>
      <DropdownItem onClick={() => setAutoConnect(server, !isAutoConnect)}>
        <FontAwesomeIcon icon={autoConnectIcon} fixedWidth /> {isAutoConnect ? 'Do not a' : 'A'}uto-connect
      </DropdownItem>
      <DropdownItem divider tag="hr" />
      <DropdownItem className="dropdown-item--danger" onClick={showModal}>
        <FontAwesomeIcon icon={deleteIcon} fixedWidth /> Remove server
      </DropdownItem>

      <DeleteServerModal redirectHome={false} server={server} isOpen={isModalOpen} toggle={hideModal} />
    </RowDropdownBtn>
  );
};

export const ManageServersRowDropdownFactory = componentFactory(ManageServersRowDropdown, ['DeleteServerModal']);
