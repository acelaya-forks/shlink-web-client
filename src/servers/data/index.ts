import type { SemVer } from '../../utils/helpers/version';

export interface ServerData {
  name: string;
  url: string;
  apiKey: string;
  forwardCredentials?: boolean;
}

export interface ServerWithId extends ServerData {
  id: string;
  autoConnect?: boolean;
}

export interface ReachableServer extends ServerWithId {
  version: SemVer;
  printableVersion: string;
}

export interface NonReachableServer extends ServerWithId {
  serverNotReachable: true;
}

export interface NotFoundServer {
  serverNotFound: true;
}

export type RegularServer = ReachableServer | NonReachableServer;

export type SelectedServer = RegularServer | NotFoundServer | null;

export type ServersMap = Record<string, ServerWithId>;

export const hasServerData = (server: SelectedServer | ServerData): server is ServerData =>
  !!(server as ServerData)?.url && !!(server as ServerData)?.apiKey;

export const isServerWithId = (server: SelectedServer | ServerWithId): server is ServerWithId =>
  !!(server as ServerWithId)?.id;

export const isReachableServer = (server: SelectedServer): server is ReachableServer =>
  !!(server as ReachableServer)?.version;

export const isNotFoundServer = (server: SelectedServer): server is NotFoundServer =>
  !!(server as NotFoundServer)?.serverNotFound;

export const getServerId = (server: SelectedServer) => (isServerWithId(server) ? server.id : '');

/**
 * Expose values that represent provided server, in a way that can be serialized in JSON or CSV strings.
 */
export const serializeServer = ({ name, url, apiKey, forwardCredentials }: ServerData): Record<string, string> => ({
  name,
  url,
  apiKey,
  forwardCredentials: forwardCredentials ? 'true' : 'false',
});

const validateServerData = (server: any): server is ServerData =>
  typeof server.url === 'string' && typeof server.apiKey === 'string' && typeof server.name === 'string';

/**
 * Provided a record, it picks the right properties to build a ServerData object.
 * @throws Error If any of the required ServerData properties is missing.
 */
export const deserializeServer = (potentialServer: Record<string, unknown>): ServerData => {
  const { forwardCredentials, ...serverData } = potentialServer;
  if (!validateServerData(serverData)) {
    throw new Error('Server is missing required "url", "apiKey" and/or "name" properties');
  }

  return {
    ...serverData,
    forwardCredentials: forwardCredentials === 'true',
  };
};
