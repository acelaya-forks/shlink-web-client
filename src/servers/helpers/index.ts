import { groupBy } from '@shlinkio/data-manipulation';
import type { ServerData, ServersMap, ServerWithId } from '../data';

/**
 * Builds a potentially unique ID for a server, based on concatenating their name and the hostname of their domain, all
 * in lowercase and replacing invalid URL characters with hyphens.
 */
function idForServer(server: ServerData): string {
  let urlSegment = server.url;
  try {
    const { host, pathname } = new URL(urlSegment);
    urlSegment = host;

    // Remove leading slash from pathname
    const normalizedPathname = pathname.substring(1);

    // Include pathname in the ID, if not empty
    if (normalizedPathname.length > 0) {
      urlSegment = `${urlSegment} ${normalizedPathname}`;
    }
  } catch {
    // If the server URL is not valid, use the value as is
  }

  return `${server.name} ${urlSegment}`.toLowerCase().replace(/[^a-zA-Z0-9-_.~]/g, '-');
}

export function serversListToMap(servers: ServerWithId[]): ServersMap {
  const serversMap: ServersMap = {};
  servers.forEach((server) => {
    serversMap[server.id] = server;
  });

  return serversMap;
}

const serversInclude = (serversList: ServerData[], { url, apiKey }: ServerData) =>
  serversList.some((server) => server.url === url && server.apiKey === apiKey);

export type DedupServersResult = {
  /** Servers which already exist in the reference list */
  duplicatedServers: ServerData[];
  /** Servers which are new based on a reference list */
  newServers: ServerData[];
};

/**
 * Given a list of new servers, checks which of them already exist in a servers map, and which don't
 */
export function dedupServers(servers: ServersMap, serversToAdd: ServerData[]): DedupServersResult {
  const serversList = Object.values(servers);
  const { duplicatedServers = [], newServers = [] } = groupBy(
    serversToAdd,
    (server) => serversInclude(serversList, server) ? 'duplicatedServers' : 'newServers',
  );

  return { duplicatedServers, newServers };
}

/**
 * Given a servers map and a list of servers, return the same list of servers but all with an ID, ensuring the ID is
 * unique both among all those servers and existing ones
 */
export function ensureUniqueIds(existingServers: ServersMap, serversList: ServerData[]): ServerWithId[] {
  const existingIds = new Set(Object.keys(existingServers));
  const serversWithId: ServerWithId[] = [];

  serversList.forEach((server) => {
    const baseId = idForServer(server);

    let id = baseId;
    let iterations = 1;
    while (existingIds.has(id)) {
      id = `${baseId}-${iterations}`;
      iterations++;
    }

    serversWithId.push({ ...server, id });

    // Add this server's ID to the list, so that it is taken into consideration for the next ones
    existingIds.add(id);
  });

  return serversWithId;
}
