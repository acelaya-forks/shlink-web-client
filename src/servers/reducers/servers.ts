import type { PayloadAction } from '@reduxjs/toolkit';
import { createSlice } from '@reduxjs/toolkit';
import type { ServerData, ServersMap, ServerWithId } from '../data';

interface EditServer {
  serverId: string;
  serverData: Partial<ServerData>;
}

interface SetAutoConnect {
  serverId: string;
  autoConnect: boolean;
}

const initialState: ServersMap = {};

const serverWithId = (server: ServerWithId | ServerData): ServerWithId => {
  if ('id' in server) {
    return server;
  }

  return { ...server, id: crypto.randomUUID() };
};

const serversListToMap = (servers: ServerWithId[]): ServersMap => servers.reduce<ServersMap>(
  (acc, server) => ({ ...acc, [server.id]: server }),
  {},
);

export const { actions, reducer } = createSlice({
  name: 'shlink/servers',
  initialState,
  reducers: {
    editServer: {
      prepare: (serverId: string, serverData: Partial<ServerData>) => ({
        payload: { serverId, serverData },
      }),
      reducer: (state, { payload }: PayloadAction<EditServer>) => {
        const { serverId, serverData } = payload;
        return (
          !state[serverId] ? state : { ...state, [serverId]: { ...state[serverId], ...serverData } }
        );
      },
    },
    deleteServer: (state, { payload }) => {
      delete state[payload.id];
      return state;
    },
    setAutoConnect: {
      prepare: ({ id: serverId }: ServerWithId, autoConnect: boolean) => ({
        payload: { serverId, autoConnect },
      }),
      reducer: (state, { payload }: PayloadAction<SetAutoConnect>) => {
        const { serverId, autoConnect } = payload;
        if (!state[serverId]) {
          return state;
        }

        if (!autoConnect) {
          return { ...state, [serverId]: { ...state[serverId], autoConnect } };
        }

        return Object.fromEntries(
          Object.entries(state).map(([evaluatedServerId, server]) => [
            evaluatedServerId,
            { ...server, autoConnect: evaluatedServerId === serverId },
          ]),
        );
      },
    },
    createServers: {
      prepare: (servers: ServerData[]) => {
        const payload = serversListToMap(servers.map(serverWithId));
        return { payload };
      },
      reducer: (state, { payload: newServers }: PayloadAction<ServersMap>) => ({ ...state, ...newServers }),
    },
  },
});

export const { editServer, deleteServer, setAutoConnect, createServers } = actions;

export const serversReducer = reducer;
