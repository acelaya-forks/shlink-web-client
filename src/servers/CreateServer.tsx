import type { TimeoutToggle } from '@shlinkio/shlink-frontend-kit';
import { Result, useToggle } from '@shlinkio/shlink-frontend-kit';
import type { FC } from 'react';
import { useCallback, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { Button } from 'reactstrap';
import { NoMenuLayout } from '../common/NoMenuLayout';
import type { FCWithDeps } from '../container/utils';
import { componentFactory, useDependencies } from '../container/utils';
import { useGoBack } from '../utils/helpers/hooks';
import { randomUUID } from '../utils/utils';
import type { ServerData, ServersMap, ServerWithId } from './data';
import { DuplicatedServersModal } from './helpers/DuplicatedServersModal';
import type { ImportServersBtnProps } from './helpers/ImportServersBtn';
import { ServerForm } from './helpers/ServerForm';

const SHOW_IMPORT_MSG_TIME = 4000;

type CreateServerProps = {
  createServers: (servers: ServerWithId[]) => void;
  servers: ServersMap;
};

type CreateServerDeps = {
  ImportServersBtn: FC<ImportServersBtnProps>;
  useTimeoutToggle: TimeoutToggle;
};

const ImportResult = ({ type }: { type: 'error' | 'success' }) => (
  <div className="mt-3">
    <Result type={type}>
      {type === 'success' && 'Servers properly imported. You can now select one from the list :)'}
      {type === 'error' && 'The servers could not be imported. Make sure the format is correct.'}
    </Result>
  </div>
);

const CreateServer: FCWithDeps<CreateServerProps, CreateServerDeps> = ({ servers, createServers }) => {
  const { ImportServersBtn, useTimeoutToggle } = useDependencies(CreateServer);
  const navigate = useNavigate();
  const goBack = useGoBack();
  const hasServers = !!Object.keys(servers).length;
  const [serversImported, setServersImported] = useTimeoutToggle(false, SHOW_IMPORT_MSG_TIME);
  const [errorImporting, setErrorImporting] = useTimeoutToggle(false, SHOW_IMPORT_MSG_TIME);
  const [isConfirmModalOpen, toggleConfirmModal] = useToggle();
  const [serverData, setServerData] = useState<ServerData>();
  const saveNewServer = useCallback((theServerData: ServerData) => {
    const id = randomUUID();

    createServers([{ ...theServerData, id }]);
    navigate(`/server/${id}`);
  }, [createServers, navigate]);
  const onSubmit = useCallback((newServerData: ServerData) => {
    setServerData(newServerData);

    const serverExists = Object.values(servers).some(
      ({ url, apiKey }) => newServerData.url === url && newServerData.apiKey === apiKey,
    );

    if (serverExists) {
      toggleConfirmModal();
    } else {
      saveNewServer(newServerData);
    }
  }, [saveNewServer, servers, toggleConfirmModal]);

  return (
    <NoMenuLayout>
      <ServerForm title={<h5 className="mb-0">Add new server</h5>} onSubmit={onSubmit}>
        {!hasServers && (
          <ImportServersBtn tooltipPlacement="top" onImport={setServersImported} onImportError={setErrorImporting} />
        )}
        {hasServers && <Button outline onClick={goBack}>Cancel</Button>}
        <Button outline color="primary" className="ms-2">Create server</Button>
      </ServerForm>

      {serversImported && <ImportResult type="success" />}
      {errorImporting && <ImportResult type="error" />}

      <DuplicatedServersModal
        isOpen={isConfirmModalOpen}
        duplicatedServers={serverData ? [serverData] : []}
        onDiscard={goBack}
        onSave={() => serverData && saveNewServer(serverData)}
      />
    </NoMenuLayout>
  );
};

export const CreateServerFactory = componentFactory(CreateServer, ['ImportServersBtn', 'useTimeoutToggle']);
