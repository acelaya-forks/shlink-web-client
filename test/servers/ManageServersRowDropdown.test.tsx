import { screen } from '@testing-library/react';
import type { UserEvent } from '@testing-library/user-event';
import { fromPartial } from '@total-typescript/shoehorn';
import { MemoryRouter } from 'react-router';
import type { ServerWithId } from '../../src/servers/data';
import { ManageServersRowDropdownFactory } from '../../src/servers/ManageServersRowDropdown';
import { checkAccessibility } from '../__helpers__/accessibility';
import { renderWithEvents } from '../__helpers__/setUpTest';

describe('<ManageServersRowDropdown />', () => {
  const ManageServersRowDropdown = ManageServersRowDropdownFactory(fromPartial({
    DeleteServerModal: ({ open }: { open: boolean }) => (
      <span>DeleteServerModal {open ? '[OPEN]' : '[CLOSED]'}</span>
    ),
  }));
  const setAutoConnect = vi.fn();
  const setUp = (autoConnect = false) => {
    const server = fromPartial<ServerWithId>({ id: 'abc123', autoConnect });
    return renderWithEvents(
      <MemoryRouter>
        <ManageServersRowDropdown setAutoConnect={setAutoConnect} server={server} />
      </MemoryRouter>,
    );
  };
  const toggleDropdown = (user: UserEvent) => user.click(screen.getByRole('button'));

  it.each([
    [setUp],
    [async () => {
      const { user, container } = setUp();
      await toggleDropdown(user);

      return { container };
    }],
  ])('passes a11y checks', (setUp) => checkAccessibility(setUp()));

  it('renders expected amount of dropdown items', async () => {
    const { user } = setUp();

    expect(screen.queryByRole('menu')).not.toBeInTheDocument();
    await toggleDropdown(user);
    expect(screen.getByRole('menu')).toBeInTheDocument();

    expect(screen.getAllByRole('menuitem')).toHaveLength(4);
    expect(screen.getByRole('menuitem', { name: 'Connect' })).toHaveAttribute('href', '/server/abc123');
    expect(screen.getByRole('menuitem', { name: 'Edit server' })).toHaveAttribute('href', '/server/abc123/edit');
  });

  it('allows toggling auto-connect', async () => {
    const { user } = setUp();

    expect(setAutoConnect).not.toHaveBeenCalled();
    await toggleDropdown(user);
    await user.click(screen.getByRole('menuitem', { name: 'Auto-connect' }));
    expect(setAutoConnect).toHaveBeenCalledWith(expect.objectContaining({ id: 'abc123' }), true);
  });

  it('renders deletion modal', async () => {
    const { user } = setUp();

    expect(screen.queryByText('DeleteServerModal [OPEN]')).not.toBeInTheDocument();
    expect(screen.getByText('DeleteServerModal [CLOSED]')).toBeInTheDocument();

    await toggleDropdown(user);
    await user.click(screen.getByRole('menuitem', { name: 'Remove server' }));

    expect(screen.getByText('DeleteServerModal [OPEN]')).toBeInTheDocument();
    expect(screen.queryByText('DeleteServerModal [CLOSED]')).not.toBeInTheDocument();
  });

  it.each([[true], [false]])('renders expected size and icon', (autoConnect) => {
    const { container } = setUp(autoConnect);
    expect(container).toMatchSnapshot();
  });
});
