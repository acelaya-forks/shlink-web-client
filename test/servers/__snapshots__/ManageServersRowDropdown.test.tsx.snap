// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<ManageServersRowDropdown /> > renders expected size and icon 1`] = `
<div>
  <div
    class="dropdown"
  >
    <button
      aria-expanded="false"
      aria-haspopup="true"
      aria-label="Options"
      class="dropdown-btn__toggle btn btn-primary btn-sm"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-ellipsis-vertical px-1"
        data-icon="ellipsis-vertical"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 128 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M64 360a56 56 0 1 0 0 112 56 56 0 1 0 0-112zm0-160a56 56 0 1 0 0 112 56 56 0 1 0 0-112zM120 96A56 56 0 1 0 8 96a56 56 0 1 0 112 0z"
          fill="currentColor"
        />
      </svg>
    </button>
    <div
      aria-hidden="true"
      class="w-100 dropdown-menu dropdown-menu-end"
      data-bs-popper="static"
      role="menu"
      style="min-width: 210px;"
      tabindex="-1"
    >
      <a
        class="dropdown-item"
        href="/server/abc123"
        role="menuitem"
        tabindex="0"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-plug fa-fw "
          data-icon="plug"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 384 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M96 0C78.3 0 64 14.3 64 32l0 96 64 0 0-96c0-17.7-14.3-32-32-32zM288 0c-17.7 0-32 14.3-32 32l0 96 64 0 0-96c0-17.7-14.3-32-32-32zM32 160c-17.7 0-32 14.3-32 32s14.3 32 32 32l0 32c0 77.4 55 142 128 156.8l0 67.2c0 17.7 14.3 32 32 32s32-14.3 32-32l0-67.2C297 398 352 333.4 352 256l0-32c17.7 0 32-14.3 32-32s-14.3-32-32-32L32 160z"
            fill="currentColor"
          />
        </svg>
         Connect
      </a>
      <a
        class="dropdown-item"
        href="/server/abc123/edit"
        role="menuitem"
        tabindex="0"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-pen-to-square fa-fw "
          data-icon="pen-to-square"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M471.6 21.7c-21.9-21.9-57.3-21.9-79.2 0L362.3 51.7l97.9 97.9 30.1-30.1c21.9-21.9 21.9-57.3 0-79.2L471.6 21.7zm-299.2 220c-6.1 6.1-10.8 13.6-13.5 21.9l-29.6 88.8c-2.9 8.6-.6 18.1 5.8 24.6s15.9 8.7 24.6 5.8l88.8-29.6c8.2-2.7 15.7-7.4 21.9-13.5L437.7 172.3 339.7 74.3 172.4 241.7zM96 64C43 64 0 107 0 160L0 416c0 53 43 96 96 96l256 0c53 0 96-43 96-96l0-96c0-17.7-14.3-32-32-32s-32 14.3-32 32l0 96c0 17.7-14.3 32-32 32L96 448c-17.7 0-32-14.3-32-32l0-256c0-17.7 14.3-32 32-32l96 0c17.7 0 32-14.3 32-32s-14.3-32-32-32L96 64z"
            fill="currentColor"
          />
        </svg>
         Edit server
      </a>
      <button
        class="dropdown-item"
        role="menuitem"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-ban fa-fw "
          data-icon="ban"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M367.2 412.5L99.5 144.8C77.1 176.1 64 214.5 64 256c0 106 86 192 192 192c41.5 0 79.9-13.1 111.2-35.5zm45.3-45.3C434.9 335.9 448 297.5 448 256c0-106-86-192-192-192c-41.5 0-79.9 13.1-111.2 35.5L412.5 367.2zM0 256a256 256 0 1 1 512 0A256 256 0 1 1 0 256z"
            fill="currentColor"
          />
        </svg>
         
        Do not a
        uto-connect
      </button>
      <hr
        class="dropdown-divider"
        tabindex="-1"
      />
      <button
        class="dropdown-item--danger dropdown-item"
        role="menuitem"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-circle-minus fa-fw "
          data-icon="circle-minus"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 512A256 256 0 1 0 256 0a256 256 0 1 0 0 512zM184 232l144 0c13.3 0 24 10.7 24 24s-10.7 24-24 24l-144 0c-13.3 0-24-10.7-24-24s10.7-24 24-24z"
            fill="currentColor"
          />
        </svg>
         Remove server
      </button>
      <span>
        DeleteServerModal 
        [CLOSED]
      </span>
    </div>
  </div>
</div>
`;

exports[`<ManageServersRowDropdown /> > renders expected size and icon 2`] = `
<div>
  <div
    class="dropdown"
  >
    <button
      aria-expanded="false"
      aria-haspopup="true"
      aria-label="Options"
      class="dropdown-btn__toggle btn btn-primary btn-sm"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-ellipsis-vertical px-1"
        data-icon="ellipsis-vertical"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 128 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M64 360a56 56 0 1 0 0 112 56 56 0 1 0 0-112zm0-160a56 56 0 1 0 0 112 56 56 0 1 0 0-112zM120 96A56 56 0 1 0 8 96a56 56 0 1 0 112 0z"
          fill="currentColor"
        />
      </svg>
    </button>
    <div
      aria-hidden="true"
      class="w-100 dropdown-menu dropdown-menu-end"
      data-bs-popper="static"
      role="menu"
      style="min-width: 170px;"
      tabindex="-1"
    >
      <a
        class="dropdown-item"
        href="/server/abc123"
        role="menuitem"
        tabindex="0"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-plug fa-fw "
          data-icon="plug"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 384 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M96 0C78.3 0 64 14.3 64 32l0 96 64 0 0-96c0-17.7-14.3-32-32-32zM288 0c-17.7 0-32 14.3-32 32l0 96 64 0 0-96c0-17.7-14.3-32-32-32zM32 160c-17.7 0-32 14.3-32 32s14.3 32 32 32l0 32c0 77.4 55 142 128 156.8l0 67.2c0 17.7 14.3 32 32 32s32-14.3 32-32l0-67.2C297 398 352 333.4 352 256l0-32c17.7 0 32-14.3 32-32s-14.3-32-32-32L32 160z"
            fill="currentColor"
          />
        </svg>
         Connect
      </a>
      <a
        class="dropdown-item"
        href="/server/abc123/edit"
        role="menuitem"
        tabindex="0"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-pen-to-square fa-fw "
          data-icon="pen-to-square"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M471.6 21.7c-21.9-21.9-57.3-21.9-79.2 0L362.3 51.7l97.9 97.9 30.1-30.1c21.9-21.9 21.9-57.3 0-79.2L471.6 21.7zm-299.2 220c-6.1 6.1-10.8 13.6-13.5 21.9l-29.6 88.8c-2.9 8.6-.6 18.1 5.8 24.6s15.9 8.7 24.6 5.8l88.8-29.6c8.2-2.7 15.7-7.4 21.9-13.5L437.7 172.3 339.7 74.3 172.4 241.7zM96 64C43 64 0 107 0 160L0 416c0 53 43 96 96 96l256 0c53 0 96-43 96-96l0-96c0-17.7-14.3-32-32-32s-32 14.3-32 32l0 96c0 17.7-14.3 32-32 32L96 448c-17.7 0-32-14.3-32-32l0-256c0-17.7 14.3-32 32-32l96 0c17.7 0 32-14.3 32-32s-14.3-32-32-32L96 64z"
            fill="currentColor"
          />
        </svg>
         Edit server
      </a>
      <button
        class="dropdown-item"
        role="menuitem"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-circle fa-fw "
          data-icon="circle"
          data-prefix="far"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M464 256A208 208 0 1 0 48 256a208 208 0 1 0 416 0zM0 256a256 256 0 1 1 512 0A256 256 0 1 1 0 256z"
            fill="currentColor"
          />
        </svg>
         
        A
        uto-connect
      </button>
      <hr
        class="dropdown-divider"
        tabindex="-1"
      />
      <button
        class="dropdown-item--danger dropdown-item"
        role="menuitem"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-circle-minus fa-fw "
          data-icon="circle-minus"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 512A256 256 0 1 0 256 0a256 256 0 1 0 0 512zM184 232l144 0c13.3 0 24 10.7 24 24s-10.7 24-24 24l-144 0c-13.3 0-24-10.7-24-24s10.7-24 24-24z"
            fill="currentColor"
          />
        </svg>
         Remove server
      </button>
      <span>
        DeleteServerModal 
        [CLOSED]
      </span>
    </div>
  </div>
</div>
`;
